### PR TITLE
Storage cleanup

### PR DESCRIPTION
**Pull request by jforissier**
One small fix plus a new option: `xtest --clear-storage`.

### CODE REVIEW

**Comment by jforissier** [comment]
Here is an example on QEMU. `xtest` is started and interrupted, which leaves unwanted objects behind (the Ctrl-C has to happen at the right time but it's not too hard to reproduce). Therefore the next `xtest` run reports errors. Then `xtest --clear-storage` is run, which puts back the storage in a clean state as shown by the last `xtest` command which passes.

```
$ xtest 6001
Test ID: 6001
Run test suite with level=0

TEE test application started over default TEE instance
######################################################
#
# regression+pkcs11
#
######################################################

* regression_6001 Test TEE_CreatePersistentObject
o regression_6001.1 Storage id: 00000001
^C
$ xtest 6001
Test ID: 6001
Run test suite with level=0

TEE test application started over default TEE instance
######################################################
#
# regression+pkcs11
#
######################################################

* regression_6001 Test TEE_CreatePersistentObject
o regression_6001.1 Storage id: 00000001
/home/jerome/work/optee_repo_qemu_v8/out-br/build/optee_test_ext-1.0/host/xtest/regression_6000.c:801: fs_create(&sess, file_00, 0, 0x00000002 | 0x00000004, 0, data_00, sizeof(data_00), &obj, storage_id) has an unexpected value: 0xffff0003 = TEEC_ERROR_ACCESS_CONFLICT, expected 0x0 = TEEC_SUCCESS
  regression_6001.1 FAILED
o regression_6001.2 Storage id: 80000000
/home/jerome/work/optee_repo_qemu_v8/out-br/build/optee_test_ext-1.0/host/xtest/regression_6000.c:801: fs_create(&sess, file_00, 0, 0x00000002 | 0x00000004, 0, data_00, sizeof(data_00), &obj, storage_id) has an unexpected value: 0xffff0003 = TEEC_ERROR_ACCESS_CONFLICT, expected 0x0 = TEEC_SUCCESS
  regression_6001.2 FAILED
  regression_6001 FAILED
+-----------------------------------------------------
Result of testsuite regression+pkcs11 filtered by "6001":
regression_6001.1 FAILED first error at /home/jerome/work/optee_repo_qemu_v8/out-br/build/optee_test_ext-1.0/host/xtest/regression_6000.c:801
regression_6001.2 FAILED first error at /home/jerome/work/optee_repo_qemu_v8/out-br/build/optee_test_ext-1.0/host/xtest/regression_6000.c:801
regression_6001 FAILED
+-----------------------------------------------------
8 subtests of which 2 failed
1 test case of which 1 failed
131 test cases were skipped
TEE test application done!
$ xtest --clear-storage
$ xtest 6001
Test ID: 6001
Run test suite with level=0

TEE test application started over default TEE instance
######################################################
#
# regression+pkcs11
#
######################################################

* regression_6001 Test TEE_CreatePersistentObject
o regression_6001.1 Storage id: 00000001
  regression_6001.1 OK
o regression_6001.2 Storage id: 80000000
  regression_6001.2 OK
  regression_6001 OK
+-----------------------------------------------------
Result of testsuite regression+pkcs11 filtered by "6001":
regression_6001 OK
+-----------------------------------------------------
14 subtests of which 0 failed
1 test case of which 0 failed
131 test cases were skipped
TEE test application done!
$
```

**Review by jbech-linaro — COMMENTED** [review]
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

File: host/xtest/xtest_main.c [hunk 1]
```diff
@@ -169,6 +171,8 @@ int main(int argc, char *argv[])
 #endif
 	else if (argc > 1 && !strcmp(argv[1], "--stats"))
 		return stats_runner_cmd_parser(argc - 1, &argv[1]);
+	else if (argc > 1 && !strcmp(argv[1], "--clear-storage"))
```
Comment (jenswi-linaro) [inline host/xtest/xtest_main.c:174]: argc == 2 ?

Comment (jforissier) [reply host/xtest/xtest_main.c]: Fixed

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: ta/storage/storage.c [hunk 2]
```diff
@@ -662,3 +664,73 @@ TEE_Result ta_storage_cmd_get_obj_info(uint32_t param_types,
...
+	if (!obj_id)
+		goto out;
+
+	while (true) {
+		enum_res = TEE_GetNextPersistentObject(oe, &oi, obj_id,
+						       &obj_id_sz);
+		if (enum_res == TEE_ERROR_ITEM_NOT_FOUND)
+			break;
+		if (enum_res)
+			break;
```
Comment (etienne-lms) [inline ta/storage/storage.c:703]: This case relates to an unexpected error. I think it should be reported by the TA as its storage may not have been wiped.

File: ta/storage/storage.c [hunk 3]
```diff
@@ -662,3 +664,73 @@ TEE_Result ta_storage_cmd_get_obj_info(uint32_t param_types,
...
+	if (res == TEE_ERROR_ITEM_NOT_FOUND) {
+		IMSG("No object found");
+		res = TEE_SUCCESS;
+		goto out;
+	}
+	if (res)
+		goto out;
+	obj_id = TEE_Malloc(TEE_OBJECT_ID_MAX_LEN, 0);
+	if (!obj_id)
+		goto out;
```
Comment (etienne-lms) [inline ta/storage/storage.c:695]: should set `res = TEE_ERROR_OUT_OF_MEMORY`

File: ta/storage/storage.c (hunk 2 above)
Comment (jforissier) [reply ta/storage/storage.c:706]: Code updated to set `res = enum_res;`.

**Comment by mwasilew** [comment]
Does this also fix the problem with regression 4007_ecc and 4011 leaving the generated keys behind? I'm asking as I didn't yet have chance to test.

**Comment by etienne-lms** [comment]
I don't think so. This change targets ta/storageX/ TAs, not `crypt_user_ta_uuid` regression tests 4007 and 4011 are using.
I don't think 4007/4011 use persistent objects so no cleanup required. Upon session closure, resources are released. Unless I missed somthing...

**Comment by ldts** [comment]
> @etienne-lms I believe @mwasilew is referring to #652 which is related to SE050. I agree that in theory those tests should not create anything persistent, but it seems with SE050 it is not what is happening. CC @ldts.

yes it is a bit different with the secure element and I didnt have time to propose a _fix_ for that yet. Controlling the persistent storage in NVM is done from its persistent storage mirror (ie, by importing the SE keys into the PKCS#11 TA database) or by using a tool like https://github.com/foundriesio/fio-se05x-cli.git.

Adressing the issue reported by those xtests is just not too high on my priority list since real SE05x users always interface to the element using PKCS#11.

**Pull request merged**